x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/utils.ts

```diff
@@ -193,6 +195,134 @@ export const getTupleDuplicateErrorsAndUniqueRules = (
   return [Array.from(errors.values()), Array.from(rulesAcc.values())];
 };
 
+/**
+ * Query for a saved object with a given origin id and replace the
+ * id in the provided action with the _id from the query result
+ * @param action
+ * @param esClient
+ * @returns
+ */
+export const swapActionIds = async (
+  action: Action,
+  esClient: ElasticsearchClient
+): Promise<Action | Error> => {
+  try {
+    // actions are hidden SO types so we need to query
+    // with the elasticsearch client.. unless there
+    // is a better way to do this..
+    const actionWithRequestedOriginId = await esClient.search({
+      index: '.kibana',
+      body: {
+        query: {
+          bool: {
+            filter: [
+              {
+                term: {
+                  originId: {
+                    // we are attempting to import a rule that contains an action.
+                    // That action (connector) has the old, pre-8.0 _id.
+                    // We need to swap the old, pre-8.0 _id for the new id (alias_target_id)
+                    // and replace the id of the action in the rule with the new _id
+                    value: action.id,
+                  },
+                },
+              },
+              {
+                term: {
+                  type: {
+                    value: 'action',
+                  },
+                },
+              },
+            ],
+          },
+        },
+      },
+    });
+    if (actionWithRequestedOriginId.body.hits.hits.length === 1) {
+      // id is of the form 'action:1234-5678...' so I need to split off the 'action' prefix
+      return { ...action, id: actionWithRequestedOriginId.body.hits.hits[0]._id.split(':')[1] };
+    } else if (actionWithRequestedOriginId.body.hits.hits.length > 1) {
+      return new Error(
+        `action connector with originId: ${action.id} had conflicts. Please resolve these conflicts either in the file you are attempting to upload or resolve the conflicting action connector in Kibana.`
+      );
+    }
+  } catch (exc) {
+    return action;
+  }
+  return action;
```

Comment:
If I understand correctly:

Here, you are searching for any existing actions with an originId that matches the id of the object you are importing.

* If you find exactly one match, change the ID of the object you are importing (so the existing action can be overwritten
* If you find more than one match, throw an error (and return the existing action)
* If you don't find any match, don't change anything

There are a couple of additional edge cases to consider:

1. The object you are importing might have an `originId` field (if it was exported post-8.0). In that case, the new object's "origin" is its `originId` field, _not_ its `id` field.
   In pseudo-code: `const originToSearchFor = action.originId || action.id;`
2. Say you created an action in 7.16 in the Default space, exported it, and imported it into a non-Default space. Then you upgraded to 8.0 and the action ID in the non-default space got regenerated. Now you export the action from the non-Default space and you want to import it back into the Default space. You'd expect that to overwrite the original object (that has the same "origin").
However, the existing action in the Default space wouldn't have had its `id` regenerated, and it wouldn't have an `originId` field at all. So I think you should be searching for existing objects with a matching `_id` _or_ a matching `originId`. This is how we do it in import, inside the `checkOriginConflict` method:  
https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/src/core/server/saved_objects/import/lib/check_origin_conflicts.ts#L89-L93
https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/src/core/server/saved_objects/import/lib/check_origin_conflicts.ts#L55-L57

   Unfortunately it gets a little complicated because you need to search against the object's raw document `_id`. We use a [Simple Query String](https://www.elastic.co/guide/en/elasticsearch/reference/8.0/query-dsl-simple-query-string-query.html) to do that in the code above.
3. Assuming you switch to searching with the SOClient as I strongly recommend: if you export the action from the Default space, then upgrade to 8.0, then create a new (non-Default) space and try to import the action, this query wouldn't find any matching action in the new space. So it would attempt to create an action with the same ID, and the SOClient would fail with an "unresolvable conflict" error (with `metadata: { isNotOverwritable: true }`):
https://github.com/elastic/kibana/blob/f7163878c0f291ab11d2b3daa7ec9df19786198d/src/core/server/saved_objects/service/lib/preflight_check_for_create.ts#L165-L170
You can check for this before attempting to create the object -- that's what the import code does in the `checkConflicts` module (using the underlying SOClient's `checkConflicts` API):
https://github.com/elastic/kibana/blob/f7163878c0f291ab11d2b3daa7ec9df19786198d/src/core/server/saved_objects/import/lib/check_conflicts.ts#L72-L75
In that case, you should generate a new ID for this action and preserve its origin (e.g., set its `originId` field if it doesn't already have one).

Sorry that's a lot, preserving this "make at most one copy of an object in a given space" behavior turned out to be really complicated in a world where all object IDs are globally unique 😅

Reply:
Thanks for taking the time to explain all this. I think these are important things to think about when importing rules. However, as it pertains to actions specifically, I think there is one important distinction in terms of what "importing a rule with an action" actually means that may not be clear just from looking at this code.

Our custom import rule feature only **finds** existing actions. On import of a rule, we never "create" an action connector. We are only checking to see if it exists. We are essentially just importing rules with “references” to existing action connectors. **We never create new action connectors** on import. So I think the check conflicts code will not be necessary for actions, but will for sure be necessary for rule imports as we do "create" rules on import so checking conflicts will be imperative, but that will be a different PR.

Reply:
All this just to say thanks for writing all this out succinctly and clearly. It's helped me outline a few more e2e / FTR tests I will push up to this branch after PTO which should help harden our confidence in the action connector piece of the import.

<details><summary>E2e / FTR tests to be added to this PR..</summary>

1. Export rule + action reference, created in 7.16 default space
    1. Import to default space in 8.0
    2. Import to non-default space in 8.0
2. Export rule + action reference, created in 7.16 non-default space (call it space-1234)
    1. Import to default space in 8.0
    2. Import to non-default space (space-1234) in 8.0
    3. Import to non-default space (space-4567) in 8.0
</details>

As for the actual importing of a rule (which leads to the conflicts + edge cases issues you pointed out above), I'll update this ticket https://github.com/elastic/kibana/issues/118767 with the issues you've explained here (and on our previous calls) to better outline the work that needs to be started in order for the custom rule import route to be bulletproof before 8.0 ships. The checks to make sure we are not creating conflicts when importing rules will be a separate PR from this current one. Thanks again for outlining all of this here. We will make sure the rule import includes all of the necessary logic defined in the SOM import route so that we don't cause trouble down the road.

Reply:
Ahh, thanks for clarifying! So to rephrase, you are just trying to avoid broken references. That makes sense.

OK, in that case I think the overall approach looks good. Upon a second pass I think I found a few more small issues, I'll add extra comments below.

Also: In case it slipped through the cracks, it looks like #118767 hasn't been updated yet, just FYSA